ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -189,72 +262,304 @@ def waypoints_cb(self, lane_msg):
+                0.0, self.default_accel/safety_factor)
+
+        rospy.loginfo("dist_to_tl = {}, stopping_dist = {}, state = {}"
+                      .format(dist_to_tl, stopping_distance, self.state))
+
+        # handle case where car is stopped at lights and light is red
+        if self.state == 'stopped' and dist_to_tl < self.dyn_tl_buffer:
+            self.set_stopped(self.final_waypoints_start_ptr, self.
+                             final_waypoints_start_ptr + self.lookahead_wps)
+
+        elif dist_to_tl < stopping_distance + self.dyn_tl_buffer or\
+                dist_to_tl < 20:    # added last check in to try to prevent
+                                    # flipping between accel and decel
+                                    # when approaching the light
+
+            if dist_to_tl < self.dyn_tl_buffer:
+                # small buffer from stop line - stop the car if in this area
+                rospy.loginfo("Now within tl_buffer = {}".format(self.dyn_tl_buffer))
+                self.set_stopped(self.final_waypoints_start_ptr, self.
+                                 lookahead_wps)
             else:
-                for ptr in range(start_ptr, final_end_ptr):
-                    # self.next_traffic_light_wp):
-                    mod_ptr = ptr % len(self.waypoints)
-                    chord_dist += self.distance(ptr, ptr+1)
-                    v = (1-chord_dist/distance_to_rl)*velocity
-                    if v < 0.9:
-                        v = 0.0
-                    self.waypoints[mod_ptr].twist.twist.linear.x = v
-                # end for
+                # reduce car velocity to 0.0 stoping before lights
+                recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
+                                               self.lookahead_wps,
+                                               safety_factor)
             # end if else
+        elif self.waypoints[self.final_waypoints_start_ptr].get_v() >= \
+                self.default_velocity:
+                # handle case where car is at target speed and no
+                # traffic lights within stopping distance
+            self.maintain_speed(self.final_waypoints_start_ptr, self.
+                                lookahead_wps)
         else:
-            # if self.velocity < self.default_velocity:
-            #    waypoints[self.final_waypoints_start_ptr + 5].twist.twist.\
-            #        linear.x == self.velocity:
-            start_ptr = self.final_waypoints_start_ptr
-            final_end_ptr = self.final_waypoints_start_ptr + self.lookahead_wps
-            for ptr in range(start_ptr, final_end_ptr):
-                mod_ptr = ptr % len(self.waypoints)
-                self.waypoints[mod_ptr].twist.twist.linear.x = \
-                    self.default_velocity
-                # end for
-            # end if
-        # maybe this should be called from main loop
+            if self.waypoints[self.final_waypoints_start_ptr].get_v() < 1.5:
```

Comment:
See comment on line 400 about the "1.5" value.

Reply:
I will have to think about this.